### ISSUE TITLE
Basic VR Controls

### ISSUE THREAD

**Issue by RMSteenwijk** [Nexus-Aurora/vr-unity-demo#7]
Create control scheme to allow for a user to move/teleport around with a VR headset.

**Comment by MPjeremy** [comment]
Hi! I can help to create the VR controls but may I ask what VR platform it's going to be based on? oculus? steamVR? or some other VR platforms? thanks!

**Comment by RMSteenwijk** [comment]
As many as we can get, so if you feel like contributing please fork the repo and build something, if you have any questions please  send me a message or reply here.

**Comment by Sch1nken** [comment]
Just chiming in to give some insight. 
When only targeting a single platform, SteamVR is probably the most compatible one.
It natively supports all SteamVR HMDs (HTC Vive/Pro/Cosmos, Index, Pimax HMDs) as well as Windows Mixed Reality (WMR) and Oculus Rift CV1/S as well as the Oculus Quest through the USB-C Link Cable.

The downside is it will only run with minimal effort from the user when using SteamVR. So no Oculus Store version, or Standalone Quest version or even a Windows (Microsoft) Store version.

There are also some solutions that abstract the platform away. There is VRTK (though I don't know if they already implemented SteamVR 2.0, it was a huge issue sometime ago). I don't know about Unitys built-in XR system, though that might be even better if it supports all major platforms through abstraction.

**Comment by RMSteenwijk** [comment]
@Sch1nken good input

We'll have to see what works in our favour, I think the main thing is going for as many platforms as possible but a few popular ones like the quest we might need to put some extra time into because of the loss of player base when not supported that, for now integrating a simple platform will get us started. And we can build from there

**Comment by Sch1nken** [comment]
I was just about to test the Unity XR-Plugin system. It seems though that SteamVR (using OpenVR) is not yet fully stable/supported. 
More info: https://steamcommunity.com/app/250820/discussions/7/2268069450205612646/

On the other hang, the following are supported by the XR Unity plugin:
- Magic Leap
- Oculus
- WMR
- "Unity Mock HMD" (for testing purposes I assume)

**Comment by RMSteenwijk** [comment]
Interesting, which would you choose? @Sch1nken

**Comment by DeadlyMagikarps** [comment]
OpenVR is being completely overhauled. There is no set date of release to a stable build yet. It was completely destroyed in for 2019/2020 versions so it's almost impossible to develop VR with SteamVR. SteamVR is completely based on OpenVR and VRTK was also based on that. Which also completely broke for Unity 2019. 

Steam VR 2.0 did a huge overhaul as well that broke basically everything. The way input is handled now is completely in control of the developers rather than themselves. The input system is also completely different now. 

Ignore Magic Leap. That is an AR unit.
Oculus you'll need a developers license (it's free and easy to get). That will allow installations on the quest and other headsets of Oculus and the change to submit for Oculus store.

WMR is Windows Mixed Reality. This is for the Windows Store platform which is NOT the same as SteamVR. It is a completely separate build.

Unity Mock HMD is only in-editor VR simulation movements. Don't build with this yet.

The built in Unity XR kind of works but also doesn't. SteamVR is the way to go but it's in development mode. So vive an index headsets will not work in editor mode and obviously won't build.

I am surprised that this is being built on a Beta version of Unity. Which means the versioning is going to be messed up until a full build release is out for 2020. You can't go back to previous versions in Alpha/Beta stages.

There are a lot of new features in 2019/2020 for sure. But they are not stable. Unity still recommends the 2018 LTS if you plan to ship now. But then again, I have no idea on the desired deadline. And you can always do an upgrade during the dev cycle as long as we track all changes so we can fix things if something ends up breaking due to upgrade.

**Comment by Sch1nken** [comment]
I haven't done any large scale projects, at all. Neither in VR nor Unity for that matter but I started with smaller projects back in early 2016 when I first got my OG HTC Vive. 

Also I took the time to quickly test the current preview version of the OpenVR XR Unity Plugin. It is not in a good state as of right now. Constant Editor crashes (there is an issue and an available workaround for now). My height and the general scale of things was quite wrong. And apparently it currently does not work in actual builds (i.e. outside of the editor).

From my limited point of view I would say the following:
- If you want to just prototype something for now, and you want to work stuff from the ground up, and you have any non-SteamVR Hardware to start for now, go for the XR Plugins. This is the most "unity native" solution I guess.
- If you want to get going and figure things out along the way but want some guidance, maybe some abstraction would be nice. I don't know about the future or current state of VRTK (though commits are fairly recent) but back then it was able to fully incorporate the native platform libraries to their full extend (it directly uses SteamVR if you wish, or you can tell it to use the OculusVR runtime etc).
I don't know if there are similiar projects to VRTK. I vaguely remember something called "NewtonVR" but that was quite some time ago (> 2 years iirc).

Also, frameworks like VRTK come with a lot of boilerplate already done (teleportation, grabbing and throwing stuff, door handles, levers, push buttons). But seeing as this is a large scale project that might hinder more than it would help.

TL:DR; Maybe gather some more voices (especially from more experienced and recent developers).

Cheers!

**Comment by DeadlyMagikarps** [comment]
Also note that Unity is completely changing their way of XR technologies and abstracting.

Also take note that Gear VR and Google VR is no longer supported from 2019.3 and up.

Valve is revamping a new framework as the XR SDK to develop their OpenVR Unity XR. Which is the preview and dev mode. 

We will test this for sure right now. But it will not be a pushed completion other than a test mode.

https://blogs.unity3d.com/2020/01/24/unity-xr-platform-updates/

**Comment by JRHill-Bournemouth** [comment]
Adding some love for MRTK - 
https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Input/InputProviders.html
https://github.com/Microsoft/MixedRealityToolkit-Unity

It's been quietly but actively getting developed and should support all headsets - not just WMR.

Otherwise, SteamVR just works for most cases and likely the best option to get going.

**Comment by brycekahle** [comment]
I would second the Unity XR Input System for now. It doesn't sound like our needs are super complicated in the near future, so it wouldn't be too hard to swap out for a different system when that time comes.

**Comment by DeadlyMagikarps** [comment]
> I would second the Unity XR Input System for now. It doesn't sound like our needs are super complicated in the near future, so it wouldn't be too hard to swap out for a different system when that time comes.

Ya. We'll go beta for now in OpenVR using the Unity XR Input. Everything in 2020 build will deprecate a lot of older features. Lot's of the previous input system will not work with 2020 and future.

**Comment by RMSteenwijk** [comment]
@DeadlyMagikarps has created some "basic" vr controls and merged them into develop
